compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -409,9 +409,44 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_predicate)) => {
                         let trait_predicate = bound_predicate.rebind(trait_predicate);
                         let trait_predicate = self.resolve_vars_if_possible(trait_predicate);
-                        let trait_ref = trait_predicate.to_poly_trait_ref();
+                        let (main_trait_predicate, o) = if let ty::PredicateKind::Clause(
+                            ty::ClauseKind::Trait(root_pred)
+                        ) = root_obligation.predicate.kind().skip_binder()
+                            && trait_predicate.self_ty().skip_binder()
+                                == root_pred.self_ty().peel_refs()
+                            && trait_predicate.def_id() != root_pred.def_id()
+                            && Some(root_pred.def_id()) != self.tcx.lang_items().unsize_trait()
+                        {
+                            // Let's use the root obligation as the main message, as we care about
+                            // the most general case ("X doesn't implement Pattern<'_>") over the
+                            // most specific case that happened to fail ("char doesn't implement
+                            // Fn(&mut char)").
+                            //
+                            // We rely on a few heuristics to identify cases where this root
+                            // obligation is more important than the leaf obligation:
+                            //
+                            //  - the type of the leaf predicate is (roughly) the same as the type
+                            //    from the root predicate, as a proxy for "we care about the root"
+                            //  - the leaf trait and the root trait are different, so as to avoid
+                            //    talking about `&mut T: Trait` and instead remain talking about
+                            //    `T: Trait` instead
+                            //  - the root trait is not `Unsize`, as to avoid talking about it in
+                            //    `tests/ui/coercion/coerce-issue-49593-box-never.rs`.
```

Comment:
inline these heuristics with their actual checks above, to help keep them in sync with future changes

Reply:
Will do. I prefer that style as well, I've moved away from it recently because it messes up with rustfmt.